Add a two-part Terraform starter guide

ptgott opened the pull request:
Closes #41055

Add a learning series for users who wants to get started with managing Teleport resources using Terraform. The guide takes the user through some fundamental cluster setup tasks: deploying Agents, enrolling resources, adding labels, configuring roles, and setting up SSO. The series includes two guides:

- Part One: Enrolling resources
- Part Two: Configuring RBAC

Part One of the series is based on the "Deploy Agents with Terraform" guide. This change adds instructions to the guide to label infrastructure resources so the user can access those resources with the roles configured in Part Two.

Other changes:

- Rename `agent-pool-terraform` to `terraform-starter`.
- Rename the "Dynamic Resources" guide to be more explicitly about Infrastructure as Code. Organize this section into subsections to make room for the new starter guide.
- Edit the Further Reading discussion in Part One. Move the static resource discussion to Further Reading because, otherwise, there is too much explanatory text between instructions. Also mention auto-discovery.

ptgott committed:
Add a two-part Terraform starter guide

Closes #41055

Add a learning series for users who wants to get started with managing
Teleport resources using Terraform. The guide takes the user through
some fundamental cluster setup tasks: deploying Agents, enrolling
resources, adding labels, configuring roles, and setting up SSO. The
series includes two guides:

- Part One: Enrolling resources
- Part Two: Configuring RBAC

Part One of the series is based on the "Deploy Agents with Terraform"
guide. This change adds instructions to the guide to label
infrastructure resources so the user can access those resources with the
roles configured in Part Two.

Other changes:

- Rename `agent-pool-terraform` to `terraform-starter`.
- Rename the "Dynamic Resources" guide to be more explicitly about
  Infrastructure as Code. Organize this section into subsections to make
  room for the new starter guide.
- Edit the Further Reading discussion in Part One. Move the static
  resource discussion to Further Reading because, otherwise, there is
  too much explanatory text between instructions. Also mention
  auto-discovery.

ptgott commented:
@mmcallister Just putting this on your radar for when you have time to take a look. Thanks!

mmcallister reviewed (approved):
Looks great!

hugoShaka commented on docs/pages/management/infrastructure-as-code/terraform-starter/enroll-resources.mdx:
````diff
@@ -0,0 +1,749 @@
+      fields.
+   
+   </TabItem>
+   <TabItem label="Agentless OpenSSH Server">
+
+   ```hcl
+   resource "teleport_server" "openssh_agentless" {
+     version = "v2"
+     sub_kind = "openssh"
+     metadata = {
+       name        = "example"
+       description = "Test server"
+       labels = {
+           // Teleport adds this label by default, so add it here to
+           // ensure a consistent state.
+           "teleport.dev/origin" = "dynamic"
+           // Change "dev" to "prod" as appropriate.
+           "env" = "dev"
+         }
+       }
+     spec = {
+       addr = "198.51.100.1:22"
+       hostname = "example"
+     }
+   }
+   ```
+
+   1. Update `spec.addr` with the domain name or IP address of the server as well
+      as the port. If you manage your database with Terraform, you can set it to
+      an attribute of the corresponding Terraform resource to establish a
+      dependency relationship.
+   1. Update `hostname` with the hostname of the server as shown by Teleport
+      client tools (Teleport Connect, `tsh`, and the Web UI). 
+   1. Read the [Terraform Provider
+      Reference](../../reference/terraform-provider/resources/server.mdx) for
+      the `teleport_database` resource and adjust any additional configuration
+      fields.
+
+   </TabItem>
````
I don't think those instructions are enough to successfully enroll an agentless openssh node? Agentless SSH has many prerequisites including:
- trusting the teleport CA on the openssh node
- have Teleport issue SSH certificates for openssh and load them on the node
- the node must be publicly reachable from the proxy

I think it would be better to link to not mention agentless SSH here as the subject is too complex and the guide does not provide enough information for a user to achieve a successful setup.

hugoShaka commented on docs/pages/management/infrastructure-as-code/terraform-starter/enroll-resources.mdx:
```diff
@@ -0,0 +1,749 @@
+   `agent_roles` to include the token roles that correspond to the Teleport
+   services you plan to run on your Agent nodes:
+
+   | Role value  | Service it enables           |
+   |-------------|------------------------------|
+   | `App`       | Teleport Application Service |
+   | `Discovery` | Teleport Discovery Service   |
+   | `Db`        | Teleport Database Service    |
+   | `Kube`      | Teleport Kubernetes Service  |
+   | `Node`      | Teleport SSH Service         |
```
Maybe adding a link to the introduction page of each service would help users not familiar with our services.

hugoShaka commented on docs/pages/management/infrastructure-as-code/terraform-starter/enroll-resources.mdx:
```diff
@@ -0,0 +1,749 @@
+   | `App`       | Teleport Application Service |
+   | `Discovery` | Teleport Discovery Service   |
+   | `Db`        | Teleport Database Service    |
+   | `Kube`      | Teleport Kubernetes Service  |
```
Setting up the kube service is more complex than passing the role and starting an agent.

You build an agent that dynamically servers multiple clusters like for apps and DBs but it's way harder and not supported by the Terraform provider currently.

The easiest way to start with app access is to deploy the `teleport-kube-agent` Helm chart.

I think we should not mention roles that this guide doesn't cover (or add the missing `WindowsDesktop`, `Okta`, and `MDM` roles, but I'm not sure how actionable this is)

hugoShaka commented on docs/pages/management/infrastructure-as-code/terraform-starter/enroll-resources.mdx:
```diff
@@ -0,0 +1,749 @@
+documentation section for each kind of resource you would like to enroll:
+
+- [Databases](../../../enroll-resources/database-access/database-access.mdx)
+- [Windows
+  desktops](../../../enroll-resources/desktop-access/desktop-access.mdx)
+- [Kubernetes
+  clusters](../../../enroll-resources/kubernetes-access/kubernetes-access.mdx)
```
Windows desktop and Kube clusters are not supported in the provider:
- windows dekstop, the resource is missing an abstraction (you cannot declare a windows desktop and leave the correct agent pick it up, you have to hard code the agent name)
- kube clusters, there is not technical reason, it's just not implemented because the feature was not priotitized

hugoShaka commented on docs/pages/management/infrastructure-as-code/terraform-starter/rbac.mdx:
```diff
@@ -0,0 +1,537 @@
+            claim = "group"
+            value = "developers"
+            roles = [
+              module.dev_role.role_name
```
Thank you so much for referencing the name dynamically instead of putting a string, we got this wrong in many of our own examples.

This is very important as TF needs this to understand there is a dependency between both resources. This ensures you cannot try to delete a role which is still in the mappings.

ptgott commented:
Feedback item TODOs:

- [ ] Update the Agent guide diagram to not use the old logo ([comment](https://github.com/gravitational/teleport/pull/44834#discussion_r1702376747))
- [ ] Accommodate #44954 ([comment](https://github.com/gravitational/teleport/pull/44834#discussion_r1702379137))
- [ ] Fix agentless SSH server instructions ([comment](https://github.com/gravitational/teleport/pull/44834#discussion_r1702386671))
- [ ] Fix Kubernetes Service instructions ([comment](https://github.com/gravitational/teleport/pull/44834#discussion_r1702389110))

ptgott committed:
Respond partially to hugoShaka feedback

- Fix link syntax.
- Set expectations that the child modules downloaded in the two guides
  are examples.
- Remove diagram with outdated logo.
- Use purpose-oriented names for TF files instead of `main.tf`
- Use `tctl terraform env` to get credentials

ptgott commented on docs/pages/management/infrastructure-as-code/terraform-starter/enroll-resources.mdx:
```diff
@@ -0,0 +1,749 @@
+   </TabItem>
+   </Tabs>
+
+1. Create a file called `main.tf` with the following content, which imports the
```
Part Two of the series assumes that the reader is editing the same `main.tf` as Part One to add RBAC-related child modules. Would it make sense to change that to use, say, `agent.tf` for Agent-related child modules (Part One) and `rbac.tf` for RBAC-related ones (Part Two)?

ptgott commented on docs/pages/management/infrastructure-as-code/terraform-starter/enroll-resources.mdx:
````diff
@@ -0,0 +1,749 @@
+
+   ```code
+   $ git clone --depth=1 https://github.com/gravitational/teleport teleport-clone
+   $ cp -R teleport-clone/examples/terraform-starter/agent-installation teleport
+   $ cp -R teleport-clone/examples/terraform-starter/aws cloud
````
I've added text in Step 1 of each guide

ptgott commented on docs/pages/management/infrastructure-as-code/terraform-starter/enroll-resources.mdx:
```diff
@@ -0,0 +1,749 @@
+  |`terraform-identity/identity`|Identity file used by the Terraform provider (renewed by a Machine ID Bot).|
+  |`create_and_run_terraform_bot.sh`|Script to start a Machine ID bot.|
+  |`provider.tf`|Configuration for the Teleport Terraform provider.|
+  |`remove_terraform_bot.sh`|Script to clean up the Machine ID Bot.|
```
Thanks! I've updated the two guides in the series to use `tctl terraform env`

ptgott committed:
Only deploy SSH Service in enroll-resources.mdx

Responds to hugoShaka feedback.

Resources that are not SSH servers registered via the SSH Service have
some complexities that don't make sense to cover in this guide. Instead,
for simplicity, the guide has the user deploy SSH Service instances,
then includes the possibility of enrolling other kinds of resources in a
"Next steps" section.

Changes:
- Remove the `agent_roles` input in the
  `terraform-starter/agent-installation` module.
- Only deploy a token for the Node role, and configure the Agents
  deployed by the module to only enable the SSH Service.
- Add an agent_labels input to the `agent-installation` module.
- Move non-server deployment instructions to "Next steps".

thedevelopnik commented on docs/pages/admin-guides/infrastructure-as-code/terraform-starter/rbac.mdx:
````diff
@@ -0,0 +1,549 @@
+}
+```
+
+## Step 3/4. Configure the single sign-on connector
````
Is there a good way to make this part optional? For a quick getting started guide, adding in SAML/OIDC can complicate it fast and add potential roadblocks if it doesn't go well. And as a user, I might not have access to a test/demo SSO provider to do experiments with and get stuck because of that.

thedevelopnik replied:
Overall the guide looks great!

ptgott committed:
Respond to thedevelopnik feedback

Note that you can skip the SSO step in the RBAC guide. Include links to
relevant docs.

ptgott commented:
@zmb3 Would you or someone on your team have time to complete the Engineering review of this? Thanks!

ptgott commented:
Hi @hugoShaka, I'm putting this back on your radar for when you have time to review. Thanks!

bernardjkim commented on docs/pages/admin-guides/infrastructure-as-code/terraform-starter/rbac.mdx:
```diff
@@ -0,0 +1,566 @@
+   }
+   
+   module "prod_role" {
+     source = "./env_role"
```
```suggestion
     source        = "./env_role"
```

bernardjkim commented on docs/pages/admin-guides/infrastructure-as-code/terraform-starter/rbac.mdx:
```diff
@@ -0,0 +1,566 @@
+   }
+
+   module "dev_role" {
+     source = "./env_role"
```
```suggestion
     source        = "./env_role"
```

bernardjkim commented on docs/pages/admin-guides/infrastructure-as-code/terraform-starter/rbac.mdx:
````diff
@@ -0,0 +1,566 @@
+```hcl
+module "dev_role" {
+  principals = {
+    logins = ["dev"]
````
```suggestion
    logins            = ["dev"]
```

bernardjkim commented on docs/pages/admin-guides/infrastructure-as-code/terraform-starter/rbac.mdx:
```diff
@@ -0,0 +1,566 @@
+
+module "prod_role" {
+  principals = {
+    logins = ["root"]
```
```suggestion
    logins            = ["root"]
```

bernardjkim commented on docs/pages/admin-guides/infrastructure-as-code/terraform-starter/rbac.mdx:
````diff
@@ -0,0 +1,566 @@
+   ```hcl
+        saml_attributes_to_roles = [
+          {
+            name = "group"
````
```suggestion
            name  = "group"
```

bernardjkim commented on docs/pages/admin-guides/infrastructure-as-code/terraform-starter/rbac.mdx:
```diff
@@ -0,0 +1,566 @@
+            ]
+          },
+          {
+            name = "group"
```
```suggestion
            name  = "group"
```

bernardjkim reviewed (approved):
Looks good to me. Just some terraform fmt indentation.

ptgott committed:
Respond to bernardjkim feedback
